misc/mdns/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+                .poll_send_to(&to_send, &From::from(([224, 0, 0, 251], 5353)))
+            {
+                Ok(Async::Ready(bytes_written)) => {
+                    debug_assert_eq!(bytes_written, to_send.len());
+                }
+                Ok(Async::NotReady) => {
+                    self.send_buffers.insert(0, to_send);
+                    break;
+                }
+                Err(_) => {
+                    // Errors are non-fatal because they can happen for example if we lose
+                    // connection to the network.
+                    self.send_buffers.clear();
+                    break;
+                }
+            }
+        }
+
+        // Check for any incoming packet.
+        match self.socket.poll_recv_from(&mut self.recv_buffer) {
+            Ok(Async::Ready((len, from))) => {
+                match Packet::parse(&self.recv_buffer[..len]) {
+                    Ok(packet) => {
+                        if packet.header.query {
+                            if packet
+                                .questions
+                                .iter()
+                                .any(|q| q.qname.to_string().as_bytes() == SERVICE_NAME)
+                            {
+                                return Async::Ready(MdnsPacket::Query(MdnsQuery {
+                                    from,
+                                    query_id: packet.header.id,
+                                    send_buffers: &mut self.send_buffers,
+                                }));
+                            } else if packet
+                                .questions
+                                .iter()
+                                .any(|q| q.qname.to_string().as_bytes() == META_QUERY_SERVICE)
+                            {
+                                // TODO: what if multiple questions, one with SERVICE_NAME and one with META_QUERY_SERVICE?
+                                return Async::Ready(MdnsPacket::ServiceDiscovery(
+                                    MdnsServiceDiscovery {
+                                        from,
+                                        query_id: packet.header.id,
+                                        send_buffers: &mut self.send_buffers,
+                                    },
+                                ));
+                            } else {
+                                // Note that ideally we would use a loop instead. However as of the
+                                // writing of this code non-lexical lifetimes haven't been merged
+                                // yet, and I can't manage to write this code without having borrow
+                                // issues.
+                                task::current().notify();
+                                return Async::NotReady;
+                            }
+                        } else {
+                            return Async::Ready(MdnsPacket::Response(MdnsResponse {
```

Comment:
IIUC, it is trivial for an adversary to inject itself into to the set of listen addresses of another peer just by answering queries.

Reply:
Yes, by design it's not secure.